File: FTPResponseExporter/src/main/java/com/checkbox/ftpresponseexporter/FTPResponseExporter.java
package com.checkbox.ftpresponseexporter;

import au.com.bytecode.opencsv.CSVWriter;
import com.checkbox.ftpresponseexporter.service.ResponseData;
import org.apache.commons.io.IOUtils;

import java.io.FileWriter;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


public class FTPResponseExporter {

    private List<ExportOptions> exportOptions = new ArrayList<>();
    private FTPOptions ftpOptions;

    public FTPResponseExporter(String exportOptionsFileName, String ftpOptionsFileName) throws Exception {
        Main.LOGGER.info("Setting up options...");

        List<List<String>> exportOptionsAsStringList =
                OptionsUtils.parseOptionsAsStringListFromFile(exportOptionsFileName, ExportOptions.NUMBER_OF_OPTIONS);

        for (List<String> optionsLine : exportOptionsAsStringList) {
            exportOptions.add(new ExportOptions(optionsLine));
        }

        ftpOptions = new FTPOptions(
                OptionsUtils.parseOptionsAsStringListFromFile(ftpOptionsFileName, FTPOptions.NUMBER_OF_OPTIONS).get(0));
    }

    private void writeStringResponsesToCSVFile(List<List<String>> responses, String csvFileName) throws IOException {
        assert responses != null && csvFileName != null;

        CSVWriter csvWriter = null;
        try {
            Main.LOGGER.info("Writing responses to CSV file...");
            csvWriter = new CSVWriter(new FileWriter(csvFileName));
            for (List<String> stringList : responses) {
                csvWriter.writeNext(stringList.toArray(new String[stringList.size()]));
            }
        } finally {
            IOUtils.closeQuietly(csvWriter);
        }
    }

    public void exportDataToFTP() {
        ExecutorService executor = Executors.newFixedThreadPool(10);
        for (final ExportOptions expOptions : exportOptions){
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    URL apiUrl = expOptions.getAPIUrl();
                    FTPUploader ftpUploader = null;
                    try {
                        ResponseData responseData = new ResponseData(apiUrl);
                        writeStringResponsesToCSVFile(responseData.exportResponses(expOptions), expOptions.getCsvFileName());
                        ftpUploader = new FTPUploader(ftpOptions.getHostname(), ftpOptions.getUsername(), ftpOptions.getPassword());
                        ftpUploader.uploadFile(expOptions.getCsvFileName(), ftpOptions.getDirectory());
                    } catch (Exception e) {
                        Main.LOGGER.severe(e.getMessage());
                    } finally {
                        if (ftpUploader != null) {
                            ftpUploader.disconnect();
                        }
                    }
                }
            });
        }
        executor.shutdown();
    }

}

File: C#/FTPResponseExporter/FTPResponseExporter/FTPResponseExporter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using CsvHelper;
using System.IO;
using System.Collections.Concurrent;
using ResponseExporter.service;
using FTPResponseExporter.ResponseDataService;
using FTPResponseExporter;

namespace ResponseExporter
{
    public class FtpResponseExporter
    {
        private List<ExportOptions> exportOptions = new List<ExportOptions>();
        private FTPOptions ftpOptions;


        public FtpResponseExporter(String exportOptionsFileName, String ftpOptionsFileName)
        {
            List<List<String>> exportOptionsAsStringList = OptionUtils.parseOptionsAsStringFromList(exportOptionsFileName, ExportOptions.NUMBER_OF_OPTIONS);

            foreach ( List<String> optionsLine in exportOptionsAsStringList)
            {exportOptions.Add(new ExportOptions(optionsLine));}

             ftpOptions = new FTPOptions(OptionUtils.parseOptionsAsStringFromList(ftpOptionsFileName,FTPOptions.NUMBER_OF_OPTIONS)[0]);
        }

        private void writeStringResponsesToCSVFile(IEnumerable<SimpleNameValueCollection> responses, String CSVFileName)
        {
           
            try
            {
                using (var sr = new StreamWriter(@CSVFileName))
                {
                    var writer = new CsvWriter(sr);

                    /*--------------------------------------------------------
                     * Write Header
                     * -------------------------------------------------------*/
                    for (int j = 0; j < responses.First().NameValueList.Count(); j++)
                    {
                        writer.WriteField(responses.First().NameValueList.ElementAt(j).Name);
                    }

                    writer.NextRecord();
                    
                    /*-------------------------------------------------------
                    * Write Responses
                    * ------------------------------------------------------*/
                    foreach (SimpleNameValueCollection record in responses)   
                    {
                        
                        for (int i = 0; i < record.NameValueList.Count(); i++)
                        {
                            writer.WriteField(record.NameValueList.ElementAt(i).Value);
                        }
                        writer.NextRecord();
                    }

                }

            }

            catch (Exception e)
            {
                Console.WriteLine("CSV file creation failed");            
            }
             
         }


        public void exportDatatoFTP()
        {
            foreach (ExportOptions expOptions in exportOptions)
            {
                Uri apiUrl = expOptions.getAPIUrl();
                try 
                {
                    ResponseDataServ repsponseData = new ResponseDataServ(apiUrl);
                    writeStringResponsesToCSVFile(repsponseData.exportResponses(expOptions), expOptions.getCsvFileName());
                    var ftpUploader = new FTPUploader(ftpOptions.getHostname(), ftpOptions.getUsername(), ftpOptions.getPassword(), expOptions.getCsvFileName());
                    ftpUploader.uploadFile(expOptions.getCsvFileName(), ftpOptions.getDirectory());
                }
                catch (Exception e)
                {
                    Console.WriteLine("Data Export Failed");
                }
                    
            }
        
        }
    }
}